src/client.rs

```diff
@@ -3516,82 +3516,121 @@ pub mod peer_online {
         }
     }
 
-    async fn create_online_stream() -> ResultType<FramedStream> {
-        let (rendezvous_server, _servers, _contained) =
-            crate::get_rendezvous_server(READ_TIMEOUT).await;
-        let tmp: Vec<&str> = rendezvous_server.split(":").collect();
-        if tmp.len() != 2 {
-            bail!("Invalid server address: {}", rendezvous_server);
-        }
-        let port: u16 = tmp[1].parse()?;
-        if port == 0 {
+    async fn create_online_stream(server_addr: String) -> ResultType<FramedStream> {
+        let (rendezvous_server, _servers, _contained) = match server_addr.as_str() {
+            "public" => {
+                (format!("{}:{}",hbb_common::config::RENDEZVOUS_SERVERS[0],hbb_common::config::RENDEZVOUS_PORT), vec![], false)
+            },
+            "rendezvous_server" => {
+                crate::get_rendezvous_server(READ_TIMEOUT).await
+            },
+            _ => {
+                (hbb_common::socket_client::check_port(server_addr, hbb_common::config::RENDEZVOUS_PORT), vec![], false)
+            }
+        };
+        let port;
+        let online_server;
+        if let Some(pos) = rendezvous_server.rfind(':') {
+            port = rendezvous_server[pos + 1 ..].parse().unwrap_or(0);
+            if port == 0 {
+                bail!("Invalid server address: {}", rendezvous_server);
+            }
+            online_server = format!("{}:{}", rendezvous_server[.. pos].to_owned(), port - 1);    
+        } else {
             bail!("Invalid server address: {}", rendezvous_server);
-        }
-        let online_server = format!("{}:{}", tmp[0], port - 1);
+        }    
         connect_tcp(online_server, CONNECT_TIMEOUT).await
     }
 
     async fn query_online_states_(
```

Comment:
I would suggest you create another function 
```
query_online_states_with_ip_port(...) 
```
to wrap old query_online_states_, so that we have minimum code change.